---
title: "Untitled"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Download and read input data

The data for for hands-on session comes from GEO entry: GSE37704, which is associated with the following publication:

Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "Differential analysis of gene regulation at transcript resolution with RNA-seq". Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703
The authors report on differential analysis of lung fibroblasts in response to loss of the developmental transcription factor HOXA1.

```{r}
# Import metadata and take a peak
colData = read.csv("GSE37704_metadata.csv", row.names=1)
head(colData)
```

```{r}
# Import metadata and take a peak
countData = read.csv("GSE37704_featurecounts.csv", row.names=1)
head(countData)
```

```{r}
rawcountData <- countData[,-1]
head(rawcountData)
```


```{r}
dim(rawcountData)
```

There are genes that have zero counts in all samples. Need to get rid of rows without data, aka with values of zero.

First identify the zero sum rows

Exclude these zero sum rows

```{r}
validcountData <- rawcountData[ rowSums(rawcountData) != 0,]
```


## Running DESeq2

```{r}
library(DESeq2)
```

```{r}
dds = DESeqDataSetFromMatrix(countData=validcountData,
                             colData=colData,
                             design=~condition)
dds = DESeq(dds)
```

```{r}
res = results(dds)
```

##Make a volcano plot from data

You plot log2 fold chang vs. -log of adjusted p-value


```{r}
plot(res$log2FoldChange, -log(res$padj))
```

To improve the visual of this, use the folowing functions

```{r}
# Make a color vector for all genes
mycols <- rep("gray", nrow(res) )

# Color red the genes with absolute fold change above 2
mycols[ abs(res$log2FoldChange) > 2 ] <- "red"

# Color blue those with adjusted p-value less than 0.01
#  and absolute fold change more than 2
inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

plot( res$log2FoldChange, -log(res$padj), col=mycols, xlab="Log2(FoldChange)", ylab="-Log(P-value)" )
```


To add annotation to my gene list I will install some bioconductor packages IN THE CONSOLE ONLY

```{r}
# BiocManager::install("AnnotationDbi")
# BiocManager::install("org.Hs.eg.db")
```

```{r}
library("AnnotationDbi")
library(org.Hs.eg.db)

columns(org.Hs.eg.db)
```

Let's map our ENSEMBLE gene ids to the more convention gene SYMBOL





```{r}
res$symbol <- mapIds(org.Hs.eg.db,
                     keys = row.names(res),
                     keytype = "ENSEMBL",
                     column = "SYMBOL",
                     multiVals = "first")
head(res)
```

```{r}
res$entrez = mapIds(org.Hs.eg.db,
                    keys= row.names(res),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")

res$name =   mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype= "ENSEMBL",
                    column= "GENENAME",
                    multiVals="first")
head(res, 10)

```



```{r}
write.csv(res, file = "DESeq_myresults.csv")
```


# Pathway Analysis

Here we are going to use the gage package for pathway analysis. Once we have a list of enriched pathways, we're going to use the pathview package to draw pathway diagrams, shading the molecules in the pathway by their degree of up/down-regulation.

```{r}
# Run in your R console (i.e. not your Rmarkdown doc!)
#BiocManager::install( c("pathview", "gage", "gageData") )

# For old vesrsions of R only (R < 3.5.0)!
#source("http://bioconductor.org/biocLite.R")
#biocLite( c("pathview", "gage", "gageData") )
```

```{r}
library(pathview)
library(gage)
library(gageData)

data(kegg.sets.hs)
data(sigmet.idx.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)
```


Our input will be a vector of fold change values with ENTREZ gene ids as names

```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```


```{r}
# Get the results of pathway analysis
keggres = gage(foldchanges, gsets=kegg.sets.hs)
attributes(keggres)
attributes(keggres$less)
```

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110")
```

```{r}
knitr::include_graphics("hsa04110.pathview.png")
```


To use reactome online we need a files of gnee IDs in a particular format

```{r}
sig_genes <- res[res$padj <= 0.05 & !is.na(res$padj), "symbol"]
print(paste("Total number of significant genes:", length(sig_genes)))
```

```{r}
write.table(sig_genes, file="significant_genes.txt", row.names=FALSE, col.names=FALSE, quote=FALSE)
```

